Optimization

jesus-gorronogoitia opened the pull request:
First DSpot optimization using Caching for Test Frameworks associated to cloned test method. The assumption is that the test framework associated to a test method should be the same that the one associated to its original test method, despite the amplifications performed.
For each cloned test method, a map is maintained to keep their bounds to their original test methods. When the associated test framework is requested, the cache is inquired for the test framework computed for its original test method.

First experiments shown a performance improvement in the range [42-45%].
Below, execution report with cache optimization:
`======= REPORT =======
PitMutantScoreSelector: 
The original test suite kills 13 mutants
The amplification results with 6 new tests
it kills 2 more mutants

2019-02-11 12:13:58,344 INFO eu.stamp_project.Main - Amplification succeed.
2019-02-11 12:13:58,344 INFO eu.stamp_project.Main - Elapsed time 726948 ms`

Next, execution report with original DSpot (no cache optimization) targeting DHell: 
`======= REPORT =======
PitMutantScoreSelector: 
The original test suite kills 13 mutants
The amplification results with 7 new tests
it kills 2 more mutants

2019-02-11 12:37:59,743 INFO eu.stamp_project.Main - Amplification succeed.
2019-02-11 12:37:59,743 INFO eu.stamp_project.Main - Elapsed time 1321376 ms`

In both cases, the DSpot execution line was:
`java -Xms2048m -Xmx4096m -jar $dspotJar --path-to-properties $dspotPropertiesFile -i 3 --test all -a MethodAdd:AllLiteralAmplifiersMethodGeneratorAmplifier --verbose`

In experiments I observed different DSpot outputs in terms of new tests generated. I observe this behavior despite using cache optimization or not. We need to double check that proposed optimization changes are not modifying DSpot behavior and output.

danglotb commented:
Hello @jesus-gorronogoitia 

The changes seems good to me but you forgot to add `eu.stamp_project.DSpotCache` class to your changes.

Also, the last pull request that I merged created a conflict and I'm sorry.

However, you should be able to resolve it easily IHMO:

```
git checkout master
git pull origin master
git checkout optimization
git rebase master
```
Resolve the conflict, then:
```
git rebase --continue
```
and push your changes by forcing:
```
git push -f <your_remote> optimization.
```

For the non-determinism of the output, I'm gonna investigate. Thank you very much, this is promising! :-)

jesus-gorronogoitia commented:
Hi  @danglotb You are right, I will fix it asap.

jesus-gorronogoitia commented:
Hi @danglotb 
I fixed the conflict and my repo is clean. I've also configured the cache size using a property I added to the general DSpot configuration, using the default value. Not yet added to the user configuration. I think I should do more experiments to see the cache size impact before suggesting the user to increase/decrease the default value (10000)

jesus-gorronogoitia commented:
Next, I will profile again Dspot, using Netbeans profiler to see the impact of cached on detected bottleneck and to investigate other computation bottlenects. I will also take some memory snapshots to investigate the memory usage using some specialized tools. I will keep you informed.

danglotb commented:
Okay @jesus-gorronogoitia 

Your plan seems to be good for me. 

But, I strongly suggest that we do one optimization at the time with one pull request. We first validate this first caching mechanism and check that there is no modification in the global output of DSpot, then we will merge this one.

jesus-gorronogoitia commented:
Hi @danglotb Sure, I agree. This PR is the first one. I will conduct more experiments to check that output is equal or equivalent (see that DSpot output is not always the same) to the one produced by Dspot without optimizations.
BTW, the PR still fails in Trevis CI. Can you tell me the reason, I compiles well for me in local, using mvn clean package -DskipTests

jesus-gorronogoitia commented:
I see 3 test failed in Trevis CI execution. I am checking building and running tests locally.

jesus-gorronogoitia commented:
I am investigating the failing test to understand the side effects of caching on them.

jesus-gorronogoitia commented:
I found the reason tests are failing. Methods are cached by name, so when methods names are overlapping returned Test frameworks could be wrong. I will use the method qualified name (full java qualified name as key) to solve it.

jesus-gorronogoitia commented:
I've fixed the failing tests by changing the key in cache for test frameworks. Now I am using the full qualified Java name of method. Now all tests pass. However, I can't build DSpot locally if I pass the tests, only skipping them. I will report it in another issue

jesus-gorronogoitia commented:
Bonjour @danglotb I see TravisCI failed execution because some errors related to style:
`error: Line has trailing spaces`
What is the Maven command line I should execute to check the style and fix it?

jesus-gorronogoitia commented:
Hi  @danglotb  Meantime I wait for your answer and we together fix this PR, I will profile again DSpot execution using Netbeans Profiler to find new bottlenecks. I will also profile memory to identify possible improvements.

danglotb commented:
You can run ` mvn checkstyle:checkstyle` to verify that the style is correct.

jesus-gorronogoitia commented:
I've fixed the code style. I should have done something wrong when fetching the master branch from upstream to my origin and then rebasing master in optimization branch, since when pushing this branch I was force to pull it, getting duplicate methods in the code I modify. I will be more careful next time.

danglotb commented on dspot/src/main/java/eu/stamp_project/utils/TypeUtils.java:
```diff
@@ -93,4 +97,25 @@ public static boolean isString(CtTypeReference type) {
+	 * @param method
+	 * @return qualified name
+	 */
+	public static String getQualifiedName (CtMethod method) {
```
You can write something like: 

```java
StringBuilder qualifiedName = new StringBuilder();
		if (method != null) {
		    if (method.getParent(CtClass.class) != null) {
		        qualifiedName.append(method.getParent(CtClass.class).getQualifiedName())
                        .append(".");
            }
            qualifiedName.append(method.getSimpleName());
        }
        return qualifiedName.toString();
```
`method.getParent(CtClass.class)` will go throught the AST using the `getParent()` accessor until it found a CtClass.
`getQualifiedName()` returns the full qualified name of a class, _e.g._ `my.package.MyClass`.

jesus-gorronogoitia replied:
Thanks for the advice. I adopted the getParent (class) signature, but the getQualifiedName() applied for CtClass does not give the qualified name as you affirmed.

danglotb replied:
That's quite strange... I'm gonna have look at this.

danglotb commented on dspot/src/main/java/eu/stamp_project/utils/DSpotCache.java:
```diff
@@ -0,0 +1,35 @@
+
+		frameworkCache = cacheManager.getCache("frameworkCache", String.class, TestFrameworkSupport.class);
+
+		// frameworkCache = cacheManager.createCache("methodCache",
```
Could remove these comments?

jesus-gorronogoitia replied:
Removed, I've kept them as a code snippet for creating additional caches, but as I guess you want to have a clean code (and I have this code reference in a bookmark) there is not problem to remove them.

danglotb commented:
In fact, something weird happened with my recent changes in DSpot-prettifier. Anyway, it is not a big deal.

Except for the two comments, I'm good with the changes. Thank you very much this is very valuable for STAMP and DSpot :).

Good job.

danglotb commented:
Merged, thank you!
